### ISSUE TITLE
Docker-compose returns null error code (illegal instruction: 4) when installing on older Mac

### ISSUE THREAD

**Issue by WidgetsBurritos** [kalabox/kalabox#901]
I just completely wiped an unused Mac Mini to use as a local dev server, since my laptop was already starving for resources, and was hoping to setup kalabox for this. I have absolutely nothing on it, apart from what comes with OS X Yosemite (I haven't updated to El Capitan yet, so I dunno if that would affect anything or not).

This is my first time attempting to use kalabox, so I apologize if I'm missing something obvious, but it keeps dying on the "Pulling images" step. 

Here's the non-verbose error:
```
info: INSTALL =>  -- Step 14/18 --
info: INSTALL =>  Pulling images...
info: COMPOSE =>  0=--project-name, 1=kalabox, 2=--file, 3=/Users/davids/.kalabox/downloads/kalabox/kalabox-1.yml, 4=pull, mode=collect
info: DOCKER-COMPOSE =>  Run exited with code: null
info: COMPOSE =>  0=--project-name, 1=kalabox, 2=--file, 3=/Users/davids/.kalabox/downloads/kalabox/kalabox-2.yml, 4=pull, mode=collect
info: DOCKER-COMPOSE =>  Run exited with code: null
info: COMPOSE =>  0=--project-name, 1=kalabox, 2=--file, 3=/Users/davids/.kalabox/downloads/kalabox/kalabox-3.yml, 4=pull, mode=collect
info: DOCKER-COMPOSE =>  Run exited with code: null
info: COMPOSE =>  0=--project-name, 1=kalabox, 2=--file, 3=/Users/davids/.kalabox/downloads/kalabox/kalabox-4.yml, 4=pull, mode=collect
info: DOCKER-COMPOSE =>  Run exited with code: null
info: COMPOSE =>  0=--project-name, 1=kalabox, 2=--file, 3=/Users/davids/.kalabox/downloads/kalabox/kalabox-5.yml, 4=pull, mode=collect
info: DOCKER-COMPOSE =>  Run exited with code: null
info: INSTALL =>  VError: Failed after 5 retries. {"max":5,"backoff":500}: codenullerr:more:
info: INSTALL =>  -- INSTALL FAILURE. --
```
Since I'm not very familiar with Kalabox, I'm not entirely sure how those yaml files are intended to work, but I can confirm they all exist and contain the exact same data:

```
proxy:
  image: 'kalabox/proxy:$KALABOX_IMG_VERSION'
  environment:
    DOMAIN: $KALABOX_DOMAIN
  ports:
    - '80:80'
    - '443:443'
    - '8160:8160'
  restart: always
dns:
  image: 'kalabox/dns:$KALABOX_IMG_VERSION'
  environment:
    IP: $KALABOX_ENGINE_IP
    DOMAIN: $KALABOX_DOMAIN
  ports:
    - '$KALABOX_ENGINE_IP:53:53/udp'
  restart: always
```
I had allowed the kalabox install script, to install all the prerequisite software, so it should all be configured to the defaults. Is there anything obvious I am missing? 

[I've attached the verbose update log](https://github.com/kalabox/kalabox/files/107416/log.txt)

Thanks for your help.

**Comment by pirog** [comment]
Weird that you are getting a `null` error code. 

Can you confirm that `docker-compose` exists at `/Users/davids/.kalabox/bin/docker-compose` and is executable?

**Comment by WidgetsBurritos** [comment]
Yup.

```
Davids-Mac-mini:bin davids$ cd ~/.kalabox/bin/
Davids-Mac-mini:bin davids$ ls -lA
total 107776
-rwxr-xr-x  1 davids  staff   4938393 Jan 27 15:20 docker-compose
-rwxr-xr-x  1 davids  staff  37990352 Jan 27 15:20 docker-machine
-rwxr-xr-x  1 davids  staff  12249300 Jan 27 15:19 syncthing
```

**Comment by pirog** [comment]
If you run `kbox update` do you get the same failures?

**Comment by WidgetsBurritos** [comment]
Yes.  That's actually how I've been replicating the process because every other command keeps telling me to run "kbox update"

```
###############################################################
                                                               
               Hi there! Updates are required!                 
                                                               
 We see you just updated your kalabox code! That means that    
 you should run our updater before you proceed.                
                                                               
          Please run `kbox update` before you proceed!                                   
                                                               
###############################################################
```

**Comment by pirog** [comment]
You have binary version 0.11.1?

I have a fresh boot of 10.10. So let me try to replicate.

**Comment by WidgetsBurritos** [comment]
Yes it's whatever the latest binary available is.  I can't even type in "kbox version" without getting the message above.

**Comment by WidgetsBurritos** [comment]
I just double checked the download history.  That is the version I'm using.

**Comment by pirog** [comment]
K. i am attempted to provision this on a clean 10.10 right now.

**Comment by WidgetsBurritos** [comment]
Okay great. While we wait, is that step trying to download remote files somewhere?  If that's the case, it could potentially be a firewall issue on my end.

**Comment by pirog** [comment]
it is trying to connect to the dockerhub to pull images down. Normally it will give you a "cannot connect to docker" sort of error. Do you know the details of your firewall?

**Comment by WidgetsBurritos** [comment]
No idea.  I'm gonna try tethering the wifi to my phone instead of my office.  We'll see if that changes anything.

**Comment by pirog** [comment]
Def let me know how that goes. I'm able to install with success on a clean OSX 10.10 build so that, in combination with where this is failing makes me think you might be right re: firewall.

**Comment by WidgetsBurritos** [comment]
Nope.  Still get the same error when connected via my phone.

**Comment by pirog** [comment]
Hmmm, yeah this just installed clean for me on 10.10. Can you try this out and see what you get

```
cd ~/.kalabox/bin
./docker-machine ssh Kalabox2
# This will drop you into the VM
docker pull kalabox/proxy
```

**Comment by WidgetsBurritos** [comment]
```
Davids-Mac-mini:kalabox davids$ cd ~/.kalabox/bin/
Davids-Mac-mini:bin davids$ ./docker-machine ssh Kalabox2
                        ##         .
                  ## ## ##        ==
               ## ## ## ## ##    ===
           /"""""""""""""""""\___/ ===
      ~~~ {~~ ~~~~ ~~~ ~~~~ ~~~ ~ /  ===- ~~~
           \______ o           __/
             \    \         __/
              \____\_______/
 _                 _   ____     _            _
| |__   ___   ___ | |_|___ \ __| | ___   ___| | _____ _ __
| '_ \ / _ \ / _ \| __| __) / _` |/ _ \ / __| |/ / _ \ '__|
| |_) | (_) | (_) | |_ / __/ (_| | (_) | (__|   <  __/ |
|_.__/ \___/ \___/ \__|_____\__,_|\___/ \___|_|\_\___|_|
Boot2Docker version 1.9.1, build master : 2bb9392 - Tue Dec 15 18:52:45 UTC 2015
Docker version 1.9.1, build a34a1d5
docker@Kalabox2:~$ 
```

**Comment by pirog** [comment]
now try to run `docker pull kalabox/proxy:dev`

**Comment by WidgetsBurritos** [comment]
```
docker@Kalabox2:~$ docker pull kalabox/proxy:dev
dev: Pulling from kalabox/proxy
ab7e84202862: Pull complete 
a923bf4b2f94: Pull complete 
ce7da8b1a0f2: Pull complete 
43d1b1c0ce1c: Pull complete 
8846a54d97d4: Pull complete 
b81b71e6def3: Pull complete 
7049f6c7758e: Pull complete 
e346a6c112e9: Pull complete 
2477e73f0347: Pull complete 
32c28bf5ee60: Pull complete 
380f091c347d: Pull complete 
Digest: sha256:104b78f8361fe253ed4d5a6c61326e915a034642ede67a259b4be06ce2a77a5a
Status: Downloaded newer image for kalabox/proxy:dev
docker@Kalabox2:~$ 
```

**Comment by pirog** [comment]
OK so you are able to connect directly ergo probably not a networking issue.

**Comment by WidgetsBurritos** [comment]
I never installed any of Apple developer tools. You think that would have anything to do with it? I can install that really quick and test it.

**Comment by pirog** [comment]
Ok. `exit` out of the VM and give this a try

```
# should be in ~/.kalabox/bin
eval $(./docker-machine env Kalabox2)
export KALABOX_IMG_VERSION=dev 
./docker-compose -f ~/.kalabox/downloads/kalabox/kalabox-1.yml pull
```

**Comment by pirog** [comment]
@WidgetsBurritos worth trying but pretty sure the test i just ran didn't have them installed either

**Comment by WidgetsBurritos** [comment]
```
Davids-Mac-mini:bin davids$ pwd
/Users/davids/.kalabox/bin
Davids-Mac-mini:bin davids$ eval $(./docker-machine env Kalabox2)
Davids-Mac-mini:bin davids$ export KALABOX_IMG_VERSION=dev 
Davids-Mac-mini:bin davids$ ./docker-compose -f ~/.kalabox/downloads/kalabox/kalabox-1.yml pull
Illegal instruction: 4
Davids-Mac-mini:bin davids$ 
```

**Comment by pirog** [comment]
ok sweet. there's our error.

**Comment by pirog** [comment]
https://github.com/docker/compose/issues/1885

**Comment by WidgetsBurritos** [comment]
Okay cool.  Let me check that out and I'll let you know how that goes.

**Comment by pirog** [comment]
If you install compose you need to make the binary ends up replacing the one in `~/.kalabox/bin/` we are using compose 1.5.2

**Comment by WidgetsBurritos** [comment]
Just to make sure I'm understanding you properly, you're saying replace ~/.kalabox/bin/docker-compose with the new version of compose I install?

**Comment by pirog** [comment]
yes. there actually seems to be a binary available here which claims to resolve the issue
https://dl.bintray.com/docker-compose/master/docker-compose-Darwin-x86_64

**Comment by WidgetsBurritos** [comment]
Well I just went ahead and just installed home-brew and pip as suggested in that other issue and installed it that way.  It seems to have worked. 

```
Davids-Mac-mini:bin davids$ mv docker-compose docker-compose.old
Davids-Mac-mini:bin davids$ ln -s /usr/local/bin/docker-compose ./docker-compose
Davids-Mac-mini:bin davids$ ls -lA
total 107784
lrwxr-xr-x  1 davids  staff        29 Jan 27 16:45 docker-compose -> /usr/local/bin/docker-compose
-rwxr-xr-x  1 davids  staff   4938393 Jan 27 15:20 docker-compose.old
-rwxr-xr-x  1 davids  staff  37990352 Jan 27 15:20 docker-machine
-rwxr-xr-x  1 davids  staff  12249300 Jan 27 15:19 syncthing
Davids-Mac-mini:bin davids$ kbox update

info: INSTALL =>  -- Step 14/18 --
info: INSTALL =>  Pulling images...
info: COMPOSE =>  0=--project-name, 1=kalabox, 2=--file, 3=/Users/davids/.kalabox/downloads/kalabox/kalabox-1.yml, 4=pull, mode=collect
Pulling proxy (kalabox/proxy:v0.11.1)...
info: DOCKER-COMPOSE =>  v0.11.1: Pulling from kalabox/proxy
info: DOCKER-COMPOSE =>  Digest: sha256:c64edf8e8f8b9ab26235a4bd3e1d49802d0aa4821eba08ea85e19eb2efe2a854
Pulling dns (kalabox/dns:v0.11.1)...
info: DOCKER-COMPOSE =>  Status: Downloaded newer image for kalabox/proxy:v0.11.1
info: DOCKER-COMPOSE =>  v0.11.1: Pulling from kalabox/dns
info: DOCKER-COMPOSE =>  Digest: sha256:16e0317e4002b84d98a32d3eba7b7990de0bf5ba1052f4134239b8ffa4afa078
info: DOCKER-COMPOSE =>  Status: Downloaded newer image for kalabox/dns:v0.11.1
info: DOCKER-COMPOSE =>  Run exited with code: 0
info: COMPOSE =>  0=--project-name, 1=kalabox, 2=--file, 3=/Users/davids/.kalabox/downloads/kalabox/kalabox-2.yml, 4=pull, mode=collect
Pulling syncthing (kalabox/syncthing:v0.11.1)...
info: DOCKER-COMPOSE =>  v0.11.1: Pulling from kalabox/syncthing
info: DOCKER-COMPOSE =>  Digest: sha256:8c98288339427a7177e4767484dcad5e97eb0f67935723df40c7c4a6b92d176a
info: DOCKER-COMPOSE =>  Status: Downloaded newer image for kalabox/syncthing:v0.11.1
info: DOCKER-COMPOSE =>  Run exited with code: 0
info: COMPOSE =>  0=--project-name, 1=kalabox, 2=--file, 3=/Users/davids/.kalabox/downloads/kalabox/kalabox-3.yml, 4=build, mode=collect
proxy uses an image, skipping
dns uses an image, skipping
info: DOCKER-COMPOSE =>  Run exited with code: 0
info: COMPOSE =>  0=--project-name, 1=kalabox, 2=--file, 3=/Users/davids/.kalabox/downloads/kalabox/kalabox-4.yml, 4=build, mode=collect
syncthing uses an image, skipping
info: DOCKER-COMPOSE =>  Run exited with code: 0
info: INSTALL =>  -- OK! 78% complete! --

info: INSTALL =>  -- Step 15/18 --
info: INSTALL =>  Setting up DNS...
info: INSTALL =>  -- OK! 83% complete! --

info: INSTALL =>  -- Step 16/18 --
info: INSTALL =>  Creating services...
info: COMPOSE =>  0=--project-name, 1=kalabox, 2=--file, 3=/Users/davids/.kalabox/downloads/kalabox/kalabox-5.yml, 4=up, 5=-d, 6=--force-recreate, mode=collect
Creating kalabox_proxy_1
Creating kalabox_dns_1
info: DOCKER-COMPOSE =>  Run exited with code: 0
info: INSTALL =>  -- OK! 89% complete! --

info: INSTALL =>  -- Step 17/18 --
info: INSTALL =>  Updating install info...
info: INSTALL =>  -- OK! 94% complete! --

info: INSTALL =>  -- Step 18/18 --
info: INSTALL =>  Finishing install
info: INSTALL =>  -- OK! 100% complete! --

info: INSTALL =>  Huzzah! Victory!
```

**Comment by pirog** [comment]
Yup that looks good to me. Try hitting http://random.kbox in your browser. Should get error 400 and some clouds. Means you should be installed.

Can you detail step by step what you did to resolve this?

**Comment by WidgetsBurritos** [comment]
Got it.  Yeah it works.   And yeah, let me organize everything and i'll get you a list of steps.

**Comment by pirog** [comment]
thx! appreciate your help tracking this one down.

**Comment by WidgetsBurritos** [comment]
1. If not already present on system, install homebrew:
```
/usr/bin/ruby -e "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/master/install)"
```

2. If not already present on system, install pip:
```
pip install docker-compose
```

3. Change to the kalabox bin directory, and replace the existing docker-compose with the new docker-compose.
```
cd ~/.kalabox/bin
mv docker-compose docker-compose.old
ln -s /usr/local/bin/docker-compose ./docker-compose
```

A few notes:
- I just renamed the old docker-compose to docker-compose.old but it can probably safely be deleted. 
- I used a symbolic link instead of making a copy of docker-compose to avoid having duplicate files.  Not sure what makes more since in the context of kalabox as a whole, but it made sense for me. 
- I never attempted just downloading the updated binary you mentioned above, but I'd imagine that would work as well.

**Comment by WidgetsBurritos** [comment]
You're welcome.  I'm just glad it works now.  Thank you for your help.

**Comment by pirog** [comment]
Sure thing. I've created another issue to upgrade compose when its ready to go.
https://github.com/kalabox/kalabox/issues/902

Will close this and route to that.

**Comment by WidgetsBurritos** [comment]
ack. Just realized i messed up my steps above.  Let me fix them really quick.

**Comment by WidgetsBurritos** [comment]
Okay, I updated the steps above.

**Comment by pirog** [comment]
Looks great! Thanks again.

**Comment by WidgetsBurritos** [comment]
You're welcome.

**Comment by pirog** [comment]
Going to reopen this

**Comment by Mchampsee** [comment]
Ran into the same issue. `pip install` didn't work for me though downloading from the link @pirog sent did work for me. [Link](https://dl.bintray.com/docker-compose/master/docker-compose-Darwin-x86_64)

And thanks for all you guys do to keep kalabox running. It's been really helpful to me.

**Comment by pirog** [comment]
Good to know. Once docker-compose 1.6 is released i think we will be able to address this pretty easily.

**Comment by pirog** [comment]
@WidgetsBurritos i would try the latest 1.6 compose binary and see if that still works. 

https://github.com/docker/compose/releases

We will be rolling this into 0.11.3 so if it works for you i think we can close this issue out.

**Comment by WidgetsBurritos** [comment]
Weird.   When I download docker-compose via the curl command posted, I get "Illegal instruction: 4" still.  
```
curl -L https://github.com/docker/compose/releases/download/1.6.0/docker-compose-`uname -s`-`uname -m` > /usr/local/bin/docker-compose
chmod +x /usr/local/bin/docker-compose
```

and then:
```
Davids-Mac-mini:bin davids$ docker-compose version
Illegal instruction: 4
```

But, if I install it via pip install it works 
```
pip install docker-compose
```

And then:
```
Davids-Mac-mini:bin davids$ docker-compose version
docker-compose version 1.6.0 build d99cad6
docker-py version: 1.7.0
CPython version: 2.7.11
OpenSSL version: OpenSSL 1.0.2e 3 Dec 2015
```

I'm not exactly sure what the difference is between those, but I'll post that over on the original docker issue as well so they can hopefully get this resolved.

**Comment by pirog** [comment]
@WidgetsBurritos what about putting it in ~/.kalabox/bin? does that make a difference?

Going to reopen and put this back into the backlog

**Comment by WidgetsBurritos** [comment]
Okay so actually after looking into it more, it's an architectural problem in older macs. 
docker/toolbox#125

So I don't think this bug gets resolved just by upgrading docker-compose to the latest version. I'm guessing pip installs an alternate binary.

**Comment by pirog** [comment]
@WidgetsBurritos i def would believe that pip uses a different version. 

Wonder why they are not using the fix in the official version though.

**Comment by WidgetsBurritos** [comment]
I'm not sure.   It does work if I use that direct download you provided earlier in the issue:
https://dl.bintray.com/docker-compose/master/docker-compose-Darwin-x86_64

Although that appears to be version 1.7.0dev, not 1.6.0

**Comment by WidgetsBurritos** [comment]
Given that they've known about the issue since August and have had a lot of other releases since then, they may just be proposing it as a work around for people with older systems, and not actually building support into the official release. That's just speculation on my end though.

**Comment by pirog** [comment]
Yeah im not completely sure either. Maybe there is a difference in how they roll the official releases vs other ones. Could possibly just grab another 1.6.0 compose release and use that instead of the official one.

**Comment by WidgetsBurritos** [comment]
Yeah.  That just seems like a hassle though if you have to do that every time you update docker-compose, and does that guarantee you won't have any issues with other architectures?   Unless you can include two binaries in the future, and if the first doesn't work, failover to the second binary.

**Comment by WidgetsBurritos** [comment]
Or maybe instead of that, during the install process do a test for that before going through all the other steps, and then provide instructions on the console on how to install a working version of docker-compose.

**Comment by pirog** [comment]
I think something like this makes sense. I'm all for trying out an alternative binary straight up while we still have a smaller user base.

**Comment by WidgetsBurritos** [comment]
From what I can gather from here: https://pypi.python.org/pypi/docker-compose/1.6.0

It appears to be compiling from the source code located here: https://pypi.python.org/packages/source/d/docker-compose/docker-compose-1.6.0.tar.gz#md5=df1e773a1adf1ea2aa81130a11082df5

I will try downloading that tarball and see if I can compile myself.

**Comment by pirog** [comment]
Im thinking that the bintray and pip builds are compiled differently from the official docker ones and that is why its different... likely its just the python version being used. 

If we can find the actual binary that pip is using i would be fine using that since its compiled directly from source.

**Comment by WidgetsBurritos** [comment]
Does pip use a binary or does it just download the source directly and compile it?  

I was just able to compile a working binary doing the following:

```
curl -O https://pypi.python.org/packages/source/d/docker-compose/docker-compose-1.6.0.tar.gz#md5=df1e773a1adf1ea2aa81130a11082df5
tar xzvf docker-compose-1.6.0.tar.gz#md5\=df1e773a1adf1ea2aa81130a11082df5
cd docker-compose-1.6.0
chmod +x setup.py
./setup.py install
```

This installs a working binary to /usr/local/bin/docker-compose

**Comment by pirog** [comment]
Good question. I guess it must compile it automatically based on your
research here? Seems kind of weird for a package manager to do that but im
not super advanced on how python rolls.

**Comment by WidgetsBurritos** [comment]
Yeah me either.  I'm kinda just guessing and checking as I go. 

To throw a wrench into things, I was unable to perform the above steps on my newer laptop.   But that may just be because I'm missing the appropriate libraries on here.  I'll keep playing with it and see.

**Comment by pirog** [comment]
usually you'll need to have python and possibly lower level stuff like gcc+ and make. this is one of the reasons why its kind of weird for a package manager to rely on user-side compilation because there is no way of knowing whether the user has the deps or not.

**Comment by WidgetsBurritos** [comment]
True. Although gcc may not be necessary. From what I understand, python doesn't compile to the machine itself, but translates it into byte code for portability.  (Similar to Java).   So I think the other system requirement is an adequate version of python.   But that is not founded on any evidence, or knowledge of that actually being the case. I'm guessing based on the information I've gathered.

**Comment by WidgetsBurritos** [comment]
Okay, I think I found what we're looking for in the pip docs: https://pip.readthedocs.org/en/1.1/other-tools.html#pip-compared-to-easy-install

> pip doesn’t do everything that easy_install does. Specifically:
> 
> It cannot install from eggs. **It only installs from source.** (In the future it would be good if it could install binaries from Windows .exe or .msi – binary install on other platforms is not a priority.)
> 

So by nature of the way pip works, it currently installs directly from the source code. Which is probably why pip works, when downloading the binary directly doesn't. There are probably some environment specific issues that get solved during compilation. 

Would it be possible to do something like this during the setup process? 

1. Download the standard binary to ~/.kalabox/bin/docker-compose.

2. Run ```~/.kalabox/bin/docker-compose version``` to make sure it responds properly.  
If so, then skip the remaining steps. 

3. Otherwise, check if python is installed. If not, produce a warning and exit setup. 

4. Check if pip is installed. If so, attempt to run 

    ```pip install --install-option="--prefix=~/.kalabox/bin/" docker-compose```  

5. If pip is not installed, download source code from latest docker-compose tarball. (e.g. https://pypi.python.org/packages/source/d/docker-compose/docker-compose-1.6.0.tar.gz#md5=df1e773a1adf1ea2aa81130a11082df5)

6. Decompress tarball and then attempt to run 

   ```python setup.py install --prefix=~/.kalabox/bin/```  

7. If that fails then produce an error explaining docker-compose couldn't be installed properly and provide a link to further instructions.  (Maybe we can setup a page for that in the wiki).

**Comment by pirog** [comment]
It *is* possible but im not sure that is what we want to do. Ultimately i dont want to add a bunch of extra code to support a use case that is diminishing (older macs) or open the door for us to provide compatibility for every edge case. 

What i would prefer doing is providing a "lowest common denominator" if possible. This would be a binary compiled on an older system that hopefully works on a larger set of machines then the official binary. 

That said i don't have an older mac so you would have to post the binary and we can throw it up on our files server. You will want to make sure you are compiling against the 1.6 docker-compose tag so it is using the same source as the official binary.

**Comment by WidgetsBurritos** [comment]
That makes sense. 

So I just tested compiling it on my older machine and copying the binary over to my newer laptop and get a "bad interpreter" error. 

```
rsync davids@10.10.10.50:/usr/local/bin/docker-compose ~/tmp/docker-compose
~/tmp/docker-compose version
```

> bash: /Users/davids/tmp/docker-compose: /usr/local/opt/python/bin/python2.7: bad interpreter: No such file or directory

Pip appears to be "compiling" it specific to my python installation.  So I dunno if that's gonna work, unless there's some other way to compile the source. 

Maybe the alternate solution would be to just include the standard docker-machine and test to see if it works and if not then output a link to some page in the wiki with further steps the user can take.  That way, you're not wasting your time trying to accommodate older systems, while it's only a few extra steps for the user to get it up and running on their machine.

**Comment by WidgetsBurritos** [comment]
Oh actually, it turns out docker-compose isn't even a binary file in this context.  It's just a python script, which is why bash was looking for /usr/local/opt/python/bin/python2.7).    Reading through it really quick to see if I can figure out what's going on in there.

**Comment by WidgetsBurritos** [comment]
Okay this is what appears to be happening: 

docker-compose is a python script that imports the "compose" python script, which pip seems to setup all the necessary library info for in the system. Generally speaking it appears to just dump all the source in /usr/local/lib/python2.7/site-packages/compose/ (although this may vary from system to system, depending on user's version of python). 

So I think instead of actually compiling the source, pip is really just running the necessary scripts directly in python.  

When I try doing that on my newer machine though I get all kinds of weird errors. I can't actually successfully install docker-machine via pip on my laptop.  I'm thinking there's not a one-size-fits-all solution to this.

**Comment by pirog** [comment]
Interesting. I would check out the docker-compose repo on github and see what they do to actually compile the binary.

**Comment by WidgetsBurritos** [comment]
Will do.

**Comment by WidgetsBurritos** [comment]
I still can't seem to find the precise way they built the binary, but I may have found a work around.  Let me know what you think about this. 

```
cd /tmp
curl -O https://codeload.github.com/docker/compose/zip/1.6.0 
unzip -o 1.6.0 
rm 1.6.0
python /tmp/compose-1.6.0/setup.py install --root=~/.kalabox/ --prefix=""
rm -Rf /tmp/compose-1.6.0
```

All macs come preinstalled with a working version of python, so this should hopefully work.  Basically what the steps are doing is downloading and decompressing the 1.6.0 zip file into /tmp/compose-1.6.0/

Then it relies on the python setup script to install docker-compose into the ~/.kalabox/bin/ and then installs the necessary libraries in ~/.kalabox/lib/. The docker-compose that gets installed into ~/.kalabox/bin is not compiled, but it does work on both systems I've tested because it's building each for the particular system it is located on. 

Let me know if that works. (You'll probably want to test it on your end to make sure it still works properly).

**Comment by WidgetsBurritos** [comment]
I finally figured out how to actually compile it from my older system.  Turns out there's a /scripts/build-osx file hiding in there. I feel dumb now, after how much time I spent looking every else but inside the scripts directory. 

Test this one out and let me know if that works for you: 
https://www.dropbox.com/s/grdg3o0gpulylt1/docker-compose-Darwin-x86_64?dl=0

**Comment by pirog** [comment]
lol. I've done much worse, trust me. 

I'll drop this into my dev build and if it nominally checks out ill roll it into our dev builds for now so we can get some more testing.

**Comment by pirog** [comment]
whoop. ok. so here is a comparison of the two builds.

```bash
docker-compose version 1.6.0, build 9503aa2
docker-py version: 1.7.0
CPython version: 2.7.11
OpenSSL version: OpenSSL 1.0.2e 3 Dec 2015

PIROBOOK2:bin pirog$ ./docker-compose.official version
docker-compose version 1.6.0, build d99cad6
docker-py version: 1.7.0
CPython version: 2.7.9
OpenSSL version: OpenSSL 1.0.1j 15 Oct 2014
```

You might want to make sure you are building off the same commit as the official aka `d99cad6` that way we can minimize differences.

**Comment by pirog** [comment]
also, this is great work and really helps a ton!

**Comment by WidgetsBurritos** [comment]
Oh yeah.  It looks like they had updated the release branch since that last build.   I'm currently updating a few libraries on my system that seem to be taking their sweet time right now, but after those are done I will rebuild the binary and try again.

**Comment by pirog** [comment]
Awesome. As soon as you do that i'll try it out on our test macs and see
how it goes.

**Comment by WidgetsBurritos** [comment]
Sounds good.  Just FYI, apparently running the ./script/prepare-osx command to get all the various system libraries consistent with the official compile takes forever, so only do it if you're really wanting to sit around and wait. :-)

**Comment by pirog** [comment]
Working on kalabox full time prepares me for many situations like that :)

**Comment by WidgetsBurritos** [comment]
Fair enough.

**Comment by WidgetsBurritos** [comment]
Apparently running ./script/prepare-osx also breaks things.  Fun!

**Comment by WidgetsBurritos** [comment]
Okay it's working again.   You can use the same URL as last time to download the new binary:
https://www.dropbox.com/s/grdg3o0gpulylt1/docker-compose-Darwin-x86_64?dl=0

Are you worried about the CPython and OpenSSL versions?  Mine seem to be much newer than what they based theirs on.

**Comment by pirog** [comment]
i *think* those will be ok but i guess we will find out soon enough!

**Comment by WidgetsBurritos** [comment]
Okay cool.  Just let me know.

**Comment by pirog** [comment]
Def. I'm wrapping up another issue right now but then i'll run tests on a few macs with the binary you generated.

**Comment by WidgetsBurritos** [comment]
Okay.   It definitely works on both of my machines so there's that.  My old mac mini is at my office (which I'm about to leave) so I won't have access to it over the weekend, but if you need me to rebuild I can do so next week.

**Comment by pirog** [comment]
Your compose binary seems to work thus far! Going to roll it into dev for further testing and if all goes well we can use it in 0.11.3

**Comment by pirog** [comment]
Merged this into dev for now
https://github.com/kalabox/kalabox/commit/8222d0a96ce1965f480ef110ce8041b78cf82f09

**Comment by WidgetsBurritos** [comment]
Sounds good.  Glad it's working.  How frequently were you thinking about updating the binary? I'm just wondering going forward.  I don't mind regenerating them, I just don't want anyone waiting on me. While its still fresh in my mind I may need to document step by step how to do this in the future in case we need to do it again.

**Comment by pirog** [comment]
Since kalabox uses a seperate docker-compose binary in `~/.kalabox/bin` i dont think we will update often. 

The high level criteria that would cause us to update is if there are upstream fixes in docker that benefit kalabox. 

The biggest reason for us to update in this release was to get the new docker-compose stuff in 1.6 and to resolve this issue.

If we dont get any other serious issues that can be tracked to back to a docker-compose bug then we wont update.

**Comment by WidgetsBurritos** [comment]
Okay. Sounds good.

**Comment by pirog** [comment]
@WidgetsBurritos can you try out the new official compose 1.6.2 binary and see if it works / whether it still gives the illegal instruction warning?

**Comment by WidgetsBurritos** [comment]
No such luck. 

```
Davids-Mac-mini:kalabox-backups davids$ curl -L https://github.com/docker/compose/releases/download/1.6.2/docker-compose-`uname -s`-`uname -m` > /tmp/docker-compose
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   601    0   601    0     0    129      0 --:--:--  0:00:04 --:--:--   131
100 4808k  100 4808k    0     0   601k      0  0:00:07  0:00:07 --:--:-- 1697k
Davids-Mac-mini:kalabox-backups davids$ chmod +x /tmp/docker-compose 
Davids-Mac-mini:kalabox-backups davids$ /tmp/docker-compose 
Illegal instruction: 4
```

I'm guessing the problem is, they are still compiling it from the same machine they have been all along so it's just not getting compiled properly. Do you want a new version?

**Comment by pirog** [comment]
If you have time and want to roll a new one against the 1.6.2 tag def feel free. That said i don't think there is any specific reason we need the new version. More so just trying to preempt any hitherto unknown bugs.

**Comment by WidgetsBurritos** [comment]
I wonder if I could automate the process and just have it always store
in a dropbox folder somewhere, that way you can always pull from the
latest version. Let me look into it and I'll get back with you.

**Comment by WidgetsBurritos** [comment]
Okay, I built a little bash script that automates the entire process and uploads to this dropbox folder. 

https://www.dropbox.com/sh/zaeg2nvy1cwzh0h/AABAYLhi31DdDoW368Kqmh5ra?dl=0

I include the version number and the build.   I am just basing everything off of the release branch.  I'm assuming that's the branch you want or should I use the master branch instead?

**Comment by pirog** [comment]
very cool!

So i think what you should do is something like this so you can build off the exact commit they roll the official release against. 

```
git clone https://github.com/docker/compose.git
git checkout 1.6.2
./mybuildscript.sh
```

**Comment by WidgetsBurritos** [comment]
Well I'm including the git commands in my bash script as well.   I'm planning on storing multiple versions in my dropbox (unless storage space becomes an issue, which it isn't at this point).  

The idea is it will always build the latest version (i.e. the git commands are actually in my bash script).  So would finding the version number from CHANGES.md and then running `git checkout ${NEW_VERSION_NUMBER}` be sufficient?  Or would I need to attempt to reset the repository every single time?  

So for example, something like this: 

```
git fetch --all
git reset --hard 
VERSION=$(grep '[0-9]+\.[0-9]+\.[0-9]+' -m 1 -o CHANGES.md)
git checkout ${VERSION}
```

**Comment by pirog** [comment]
not exactly sure how you have this set up but i would pass the version directly into the script vs trying to pull it from the changelog since that format could change. 

`./build.sh 1.6.2`

and then something like 

```
VERSION=$1
git fetch --all
git reset --hard #not sure if you need this
git checkout ${VERSION}
```

**Comment by WidgetsBurritos** [comment]
I guess part of the reason for it was I was trying to just have it run nightly without me having to do anything.  That way you could always have the latest version. I do realize the format could change, so that is potentially an issue. I'm just wondering if there's a way to get that version automatically.

**Comment by pirog** [comment]
the most reliable way to get the most recent tags would be to use the github api

**Comment by WidgetsBurritos** [comment]
Okay.  They may actually have a script in there to help with that already.  I'm looking into it.

**Comment by WidgetsBurritos** [comment]
Okay, I found a way to do it.  I downloaded jsawk, which basically allows you to parse JSON from the command line, and was able to get the following command to work:

```
$ curl -s https://api.github.com/repos/docker/compose/tags | jsawk -a 'return this[0].name;'
1.6.2
```

So I guess my question is, do you happen to know if the first line in the tags will always be the latest version? If so then that should do the trick.

**Comment by WidgetsBurritos** [comment]
I still can't find confirmation but it does seem to be in a logical order, as compared to `git tag` which appears to just put things in an alphanumeric order. This seems apparent when considering the sequence of the release candidates relative to the actual release. 

I think worst case scenario, this may break down the line, but at least it should work as long as things keep working the way they currently do.  

Since it's just a local process on my end and not anything people will be using in production, I can always tweak it if need be.

**Comment by WidgetsBurritos** [comment]
@pirog Also, if you would prefer me uploading the file somewhere other than dropbox, let me know.   I'm not sure what's easiest for you.

**Comment by pirog** [comment]
If the dropbox URL is public and you arent charged for bandwidth then that works fine for me. I can also just upload it to our file server.

**Comment by WidgetsBurritos** [comment]
The files are only maybe 5MB, and I think the daily bandwidth limit is
around 10GB, so it all seems negligible to me. I doubt it's going to be
a high demand download.

**Comment by pirog** [comment]
I think we are actually going to use your version for all kalabox installs so it *might* be high demand at some point. 

either way i uploaded it to github so we dont have to burn your dropbox limitz

**Comment by WidgetsBurritos** [comment]
Hmmm.  Maybe I can setup an automatic push to github?

**Comment by WidgetsBurritos** [comment]
@pirog I've changed my automation script.  Instead of uploading to Dropbox, I now push to a separate github repository: https://github.com/WidgetsBurritos/docker-compose-old-mac/

I have it setup to push a copy of every known version right now.  Not that you need all the old versions, but I'm just doing it to be thorough in case anyone else needs it for any reason.

**Comment by WidgetsBurritos** [comment]
Okay so it stopped working for any version before 1.5.2 due to not having the correct library versions.  But I imagine that should be more than sufficient for most people's needs.

**Comment by WidgetsBurritos** [comment]
@pirog It appears this issue has been fixed in docker-compose 1.7.0-rc1
https://github.com/docker/compose/issues/271#issuecomment-201203732

I just tested it and it seems to work.  So I imagine once it's officially released you should be able to just use the stock binary.

**Comment by pirog** [comment]
That's great news!
